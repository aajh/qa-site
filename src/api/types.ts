export interface Login {
    token: string
}

export interface User {
    id: string
    username: string
}

export interface Answer {
    id: string
    author: string
    body: string
    created: string
    votes: number
    voteDirection: 1 | -1 | null
}

export interface Question {
    id: string,
    author: string,
    title: string,
    body: string,
    created: string,
    answers: Answer[]
}

export interface QuestionSummary {
    id: string,
    author: string,
    title: string,
    created: string,
}
