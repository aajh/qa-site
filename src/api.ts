import express from 'express';
import { v4 as uuidv4 } from 'uuid';

import * as api from './api-types';

const router = express.Router();

interface Question {
    id: string
    author: string,
    title: string,
    body: string,
    created: Date
}

interface Answer {
    id: string
    question: string
    author: string
    body: string
    created: Date
}

const questions: Question[] = require('../test_data/questions.json').map((q: any) => ({ ...q, created: new Date(q.created) }));
const answers: Answer[] = require('../test_data/answers.json').map((a: any) => ({ ...a, created: new Date(a.created) }));

router.get('/questions', (req, res) => {
    const jsonResponse: api.QuestionSummary[] = questions.map(
        ({ id, author, title, created }) => ({
            id,
            author,
            title,
            created: created.toISOString()
        })
    );
    res.json(jsonResponse);
});

router.get('/questions/:id', (req, res) => {
    const question = questions.find(q => q.id === req.params.id);
    if (question === undefined) {
        res.status(404).end();
    } else {
        const questionAnswers = answers
            .filter(a => a.question === req.params.id)
            .map(({ id, author, body, created }) => ({
                id, author, body, created: created.toISOString()
            }));
        const jsonResponse: api.Question = {
            ...question,
            created: question.created.toISOString(),
            answers: questionAnswers
        };
        res.json(jsonResponse);
    }
});

router.post('/questions', (req, res) => {
    if (req.body === undefined) {
        res.status(400).end();
        return;
    }

    const { author, title, body } = req.body;
    if (!author || !title || !body) {
        res.status(400).end();
        return;
    }

    const question: Question = {
        id: uuidv4(),
        author,
        title,
        body,
        created: new Date()
    };
    questions.push(question);

    const jsonResponse: api.Question = {
        ...question,
        created: question.created.toISOString(),
        answers: []
    };
    res.json(jsonResponse);
});

router.post('/questions/:id/answers', (req, res) => {
    const question = questions.find(q => q.id === req.params.id);
    if (question === undefined) {
        res.status(404).end();
        return;
    }

    if (req.body === undefined) {
        res.status(400).end();
        return;
    }

    const { author, body } = req.body;
    if (!author || !body) {
        res.status(400).end();
        return;
    }

    const answer: Answer = {
        id: uuidv4(),
        question: req.params.id,
        author,
        body,
        created: new Date(),
    };
    answers.push(answer);

    // eslint-disable-next-line no-shadow
    const jsonResponse: api.Answer = (({ id, author, body, created }) => ({
        id, author, body, created: created.toISOString()
    }))(answer);
    res.json(jsonResponse);
});

export default router;
