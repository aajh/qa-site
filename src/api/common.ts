import express from 'express';
import pg from 'pg';
import jwt from 'jsonwebtoken';

import { JWTPayload } from './types';

const connectionVariables: Record<string, string> = {
    development: 'DEV_DATABASE_URL',
    test: 'TEST_DATABASE_URL',
    production: 'DATABASE_URL'
};

export const pool = new pg.Pool({
    connectionString: process.env[connectionVariables[process.env.NODE_ENV]]
});

export function getDecodedToken(req: express.Request, res: express.Response): JWTPayload | null {
    const authorization = req.get('authorization');
    const token = authorization && authorization.toLowerCase().startsWith('bearer ')
        ? authorization.substring(7)
        : null;

    let decodedToken: JWTPayload | null = null;
    try {
        if (token) {
            decodedToken = jwt.verify(token, process.env.JWT_SECRET) as JWTPayload;
        }
    } catch (error) {
        // eslint-disable-next-line no-console
        console.info(error);
    }

    if (token === null || decodedToken === null || !decodedToken.id) {
        res.status(401).json({ error: 'token missing or invalid' });
    }

    return decodedToken;
}
