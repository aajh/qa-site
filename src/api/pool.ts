import pg from 'pg';

const connectionVariables: Record<string, string> = {
    development: 'DEV_DATABASE_URL',
    test: 'TEST_DATABASE_URL',
    production: 'DATABASE_URL'
};

const pool = new pg.Pool({
    connectionString: process.env[connectionVariables[process.env.NODE_ENV]]
});

export default pool;
