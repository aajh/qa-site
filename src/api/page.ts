import express from 'express';
import serialize from 'serialize-javascript';
import _ from 'lodash';

import { initialState } from '../slices';
import { pool } from './common';


export default async function renderPage(req: express.Request): Promise<string> {
    const state = _.cloneDeep(initialState);

    if (req.session?.userId) {
        const { userId } = req.session;
        const { rows: [{ username }] } = await pool.query(
            'SELECT username FROM users WHERE id = $1',
            [userId]
        );
        if (username) {
            state.user.user = {
                id: userId,
                username,
            };
        }
    }

    return `
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <title>QA</title>
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <link href="/main.css" rel="stylesheet">
    </head>
    <body>
        <div id="react-container"></div>
        <script>
            window.__PRELOADED_STATE__ = ${serialize(state)}
        </script>
        <script src="/bundle.js"></script>
    </body>
</html>
    `;
}
