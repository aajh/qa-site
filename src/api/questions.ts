import Router from 'express-promise-router';
import { v4 as uuidv4 } from 'uuid';
import pg from 'pg';

import * as api from './types';
import { pool, getUserId, isUuid } from './common';

const router = Router();

export async function getQuestionList(client: pg.ClientBase): Promise<api.QuestionSummary[]> {
    const { rows } = await client.query<api.QuestionSummary>(
        `SELECT questions.id, username as author, title, created
        FROM questions
        LEFT JOIN users ON users.id=questions.author_id
        ORDER BY created DESC`
    );
    return rows;
}

router.get('/', async (req, res) => {
    const client = await pool.connect();
    try {
        const jsonResponse = await getQuestionList(client);
        res.json(jsonResponse);
    } finally {
        client.release();
    }
});

export async function getQuestion(
    client: pg.ClientBase,
    questionId: string,
    userId: string | null
): Promise<api.Question | null> {
    if (!isUuid(questionId)) {
        return null;
    }

    const questionResult = await client.query(
        `SELECT questions.id, username as author, title, body, created::text
        FROM questions
        LEFT JOIN users ON users.id=questions.author_id
        WHERE questions.id = $1`,
        [questionId]
    );
    if (questionResult.rowCount === 0) {
        return null;
    }

    const { rows: questionAnswers } = await client.query(
        `SELECT answers.id, users.username as author, body, created::text, COALESCE(votes, 0) AS votes, answer_votes.direction AS vote_direction
        FROM answers
        LEFT JOIN users ON users.id=answers.author_id
        LEFT JOIN (
            SELECT answer_id, SUM(direction) AS votes
            FROM answer_votes
            GROUP BY answer_id
            ) AS votes ON answers.id=votes.answer_id
        LEFT JOIN answer_votes ON (answer_votes.answer_id=answers.id AND answer_votes.user_id=$2)
        WHERE question_id = $1
        ORDER BY created DESC`,
        [questionId, userId]
    );
    return {
        ...questionResult.rows[0],
        answers: questionAnswers
            // eslint-disable-next-line @typescript-eslint/camelcase
            .map(({ id, author, body, created, votes, vote_direction }) => ({
                id,
                author,
                body,
                created,
                votes: Number(votes),
                voteDirection: vote_direction,
            }))
    };
}

router.get('/:id', async (req, res) => {
    const client = await pool.connect();
    try {
        const userId = getUserId(req, res, false);
        const question = await getQuestion(client, req.params.id, userId);
        if (question !== null) {
            res.json(question);
        } else {
            res.status(404).json({ error: 'not found' });
        }
    } finally {
        client.release();
    }
});

router.post('/', async (req, res) => {
    const client = await pool.connect();
    try {
        if (req.body === undefined) {
            res.status(400).json({ error: 'invalid request' });
            return;
        }

        const { title, body } = req.body;
        const isInvalid = !title || !body
            || typeof title !== 'string' || title.length > 512
            || typeof body !== 'string' || body.length === 0;
        if (isInvalid) {
            res.status(400).json({ error: 'invalid request' });
            return;
        }

        const userId = getUserId(req, res);
        if (userId === null) {
            return;
        }

        const { rows } = await client.query(
            `INSERT INTO questions(id, author_id, title, body, created)
            VALUES ($1, $2, $3, $4, NOW())
            RETURNING id, title, body, created`,
            [uuidv4(), userId, title, body]
        );

        const { rows: [{ username }] } = await client.query(
            'SELECT username FROM users WHERE id = $1',
            [userId]
        );

        const jsonResponse: api.Question = {
            ...rows[0],
            author: username,
            answers: []
        };
        res.json(jsonResponse);
    } finally {
        client.release();
    }
});

router.post('/:id/answers', async (req, res) => {
    const client = await pool.connect();
    try {
        if (!isUuid(req.params.id)) {
            res.status(404).json({ error: 'not found' });
            return;
        }
        const { rows: [{ count }] } = await client.query(
            'SELECT COUNT(*) FROM questions WHERE id = $1',
            [req.params.id]
        );
        if (Number(count) === 0) {
            res.status(404).json({ error: 'not found' });
            return;
        }

        if (req.body === undefined) {
            res.status(400).json({ error: 'invalid request' });
            return;
        }

        const { body } = req.body;
        if (!body || typeof body !== 'string' || body.length === 0) {
            res.status(400).json({ error: 'invalid request' });
            return;
        }

        const userId = getUserId(req, res);
        if (userId === null) {
            return;
        }

        const { rows: [answer] } = await client.query(
            `INSERT INTO answers(id, question_id, author_id, body, created)
            VALUES ($1, $2, $3, $4, NOW())
            RETURNING id, body, created`,
            [uuidv4(), req.params.id, userId, body]
        );

        const { rows: [{ username }] } = await client.query(
            'SELECT username FROM users WHERE id = $1',
            [userId]
        );

        const jsonResponse: api.Answer = {
            ...answer,
            author: username,
            votes: 0,
        };
        res.json(jsonResponse);
    } finally {
        client.release();
    }
});

router.post('/:questionId/answers/:answerId/vote', async (req, res) => {
    const client = await pool.connect();
    try {
        if (!isUuid(req.params.questionId) || !isUuid(req.params.answerId)) {
            res.status(404).json({ error: 'not found' });
            return;
        }
        const { rows: [{ count }] } = await client.query(
            'SELECT COUNT(*) FROM answers WHERE id = $1 AND question_id = $2',
            [req.params.answerId, req.params.questionId]
        );
        if (Number(count) === 0) {
            res.status(404).json({ error: 'not found' });
            return;
        }

        if (req.body === undefined) {
            res.status(400).json({ error: 'invalid request' });
            return;
        }

        const { direction } = req.body;
        if (direction !== 1 && direction !== -1 && direction !== null) {
            res.status(400).json({ error: 'invalid request' });
            return;
        }

        const userId = getUserId(req, res);
        if (userId === null) {
            return;
        }

        if (direction !== null) {
            client.query(
                `INSERT INTO answer_votes(answer_id, user_id, direction)
                VALUES ($1, $2, $3)
                ON CONFLICT (answer_id, user_id) DO UPDATE SET direction=$3`,
                [req.params.answerId, userId, direction]
            );
        } else {
            client.query(
                'DELETE FROM answer_votes WHERE answer_id = $1 AND user_id = $2',
                [req.params.answerId, userId]
            );
        }

        res.status(200).end();
    } finally {
        client.release();
    }
});

export default router;
