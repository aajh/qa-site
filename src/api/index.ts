import express from 'express';

import questions from './questions';

const router = express.Router();

router.use('/questions', questions);

export default router;
