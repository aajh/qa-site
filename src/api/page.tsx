import express from 'express';
import Router from 'express-promise-router';
import serialize from 'serialize-javascript';
import _ from 'lodash';
import React from 'react';
import { renderToString } from 'react-dom/server';
import { Provider } from 'react-redux';
import { StaticRouter } from 'react-router';
import pg from 'pg';

import { initialState, RootState } from '../slices';
import createStore from '../store';
import App from '../components/App';
import { pool, getUserId } from './common';
import { getQuestionList, getQuestion } from './questions';

async function getInitialState(req: express.Request, client: pg.ClientBase): Promise<RootState> {
    const state = _.cloneDeep(initialState);

    if (req.session?.userId) {
        const { userId } = req.session;
        const { rows: [{ username }] } = await client.query(
            'SELECT username FROM users WHERE id = $1',
            [userId]
        );
        if (username) {
            state.user.user = {
                id: userId,
                username,
            };
        }
    }

    return state;
}

async function renderPage(req: express.Request, state: RootState): Promise<string> {
    const store = createStore(state);

    const reactHtml = renderToString(
        <Provider store={store}>
            <StaticRouter location={req.url}>
                <App />
            </StaticRouter>
        </Provider>
    );

    return `
<!DOCTYPE html>
<html>
    <head>
        <meta charset="utf-8">
        <title>QA</title>
        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <link href="/main.css" rel="stylesheet">
    </head>
    <body>
        <div id="react-container">${reactHtml}</div>
        <script>
            window.__PRELOADED_STATE__ = ${serialize(state)}
        </script>
        <script src="/bundle.js"></script>
    </body>
</html>
    `;
}

const router = Router();

async function getBasePage(req: express.Request, res: express.Response): Promise<void> {
    const client = await pool.connect();
    try {
        const state = await getInitialState(req, client);
        const page = await renderPage(req, state);
        res.send(page);
    } finally {
        client.release();
    }
}

router.get('/', async (req, res) => {
    const client = await pool.connect();
    try {
        const state = await getInitialState(req, client);

        state.questionList.questionList = await getQuestionList(client);
        state.questionList.loading = false;

        const page = await renderPage(req, state);
        res.send(page);
    } finally {
        client.release();
    }
});

router.get('/questions/ask', getBasePage);

router.get('/questions/:id', async (req, res) => {
    const client = await pool.connect();
    try {
        const state = await getInitialState(req, client);

        const userId = getUserId(req, res, false);
        const question = await getQuestion(client, req.params.id, userId);

        if (question === null) {
            res.redirect('/404');
            return;
        }

        state.question.question = question;
        state.question.loading = false;
        state.question.showExistingQuestion = true;

        const page = await renderPage(req, state);
        res.send(page);
    } finally {
        client.release();
    }
});

router.get('/*', getBasePage);

export default router;
