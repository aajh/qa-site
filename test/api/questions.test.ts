import 'isomorphic-fetch';

import { resetDatabase, SERVER_URL } from '../testHelpers';
import * as api from '../../src/api/types';

beforeEach(async () => {
    await resetDatabase();
});

test('GET /questions', async () => {
    const response = await fetch(`${SERVER_URL}/api/questions`);
    expect(response.ok).toBe(true);
    const questionList = await response.json();
    expect(questionList).toMatchSnapshot();
});

test('GET /questions/:id', async () => {
    const response = await fetch(`${SERVER_URL}/api/questions/9eae28b9-a77b-4f39-901d-140e2d4a710c`);
    expect(response.ok).toBe(true);
    const question = await response.json();
    expect(question).toMatchSnapshot();
});

test('POST /questions', async () => {
    const question = {
        title: 'Test',
        author: 'Test author',
        body: 'Test body',
    };
    const response = await fetch(`${SERVER_URL}/api/questions`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(question)
    });
    expect(response.ok).toBe(true);
    const createdQuestion: api.Question = await response.json();
    expect(createdQuestion.title).toBe(question.title);
    expect(createdQuestion.author).toBe(question.author);
    expect(createdQuestion.body).toBe(question.body);
    expect(createdQuestion.id).toBeDefined();
    expect(createdQuestion.created).toBeDefined();
    expect(createdQuestion.answers).toEqual([]);
});

test('POST /questions/:id/answers', async () => {
    const answer = {
        title: 'Test',
        author: 'Test author',
        body: 'Test body',
    };
    const response = await fetch(`${SERVER_URL}/api/questions/9eae28b9-a77b-4f39-901d-140e2d4a710c/answers`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(answer)
    });
    expect(response.ok).toBe(true);
    const createdAnswer: api.Answer = await response.json();
    expect(createdAnswer.author).toBe(answer.author);
    expect(createdAnswer.body).toBe(answer.body);
    expect(createdAnswer.id).toBeDefined();
    expect(createdAnswer.created).toBeDefined();
});
