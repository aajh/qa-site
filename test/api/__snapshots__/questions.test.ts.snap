// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET /questions 1`] = `
Array [
  Object {
    "author": "Kevin.a",
    "created": "2020-04-21T13:00:00.000Z",
    "id": "9eae28b9-a77b-4f39-901d-140e2d4a710c",
    "title": "favicon.ico 500 error when fetching robots.txt",
  },
  Object {
    "author": "Mike D",
    "created": "2020-04-05T18:16:00.000Z",
    "id": "13ebf57d-4d1d-49db-aa38-301767778c81",
    "title": "Vue recognize text pattern and replace by href to correct resource",
  },
  Object {
    "author": "donto",
    "created": "2020-04-22T07:30:00.000Z",
    "id": "96a2bec4-0c46-4d86-bce7-72dfee38e3b1",
    "title": "Sample the neighbour value from a list",
  },
]
`;

exports[`GET /questions/:id 1`] = `
Object {
  "answers": Array [
    Object {
      "author": "Michael",
      "body": "Just do something else.",
      "created": "2020-04-21T15:01:00.000Z",
      "id": "1dc42217-5ef3-40bc-ad66-fd45377359bb",
    },
    Object {
      "author": "John",
      "body": "Do this instead.",
      "created": "2020-04-22T07:16:00.000Z",
      "id": "df39ad99-64e1-4f9f-8d61-c3b95802dbf8",
    },
  ],
  "author": "Kevin.a",
  "body": "I have an error 500 on my example.com/robots.txt page.

Could this cause any SEO / crawling issues? It's just the favicon , but you never know.",
  "created": "2020-04-21T13:00:00.000Z",
  "id": "9eae28b9-a77b-4f39-901d-140e2d4a710c",
  "title": "favicon.ico 500 error when fetching robots.txt",
}
`;
